### PR TITLE
Refactor long displacement logic to MemoryReference class

### PR DESCRIPTION
**Pull request by fjeremic**
This PR has several incremental steps which are documented in individual commits. See the commit description for each individual change as to what was performed and why. The high level summary of this PR stems from some of the work done in #3113 where I've promised @0xdaryl [1] [2] that some issues identified outside the scope of the aforementioned PR will be addressed in future PRs. This is the future PR.

- Deprecate use of extended code base register
- Deprecate use of `generate(RXY|RSY|SIY)Instruction` API
- Deprecate true register model and fold all resulting dead code
- Deprecate Z codegen `startOfLiveRange` and unify ICF handling on Z
- Deprecate displacement field inside of S390 MemoryReference
- Add long and huge displacement APIs in MemoryReference
- Move `opCodeCanBeAdjustedTo` to `getLongDisplacementMnemonic` 

Once this is merged we have finally prepared the Z code generator to complete #2848.

Issue: #2848

[1] https://github.com/eclipse/omr/pull/3113#discussion_r231994002
[2] https://github.com/eclipse/omr/pull/3113#discussion_r232046867

### CODE REVIEW

**Comment by fjeremic** [comment]
Also requesting reviews from any of @dchopra001, @NigelYiboYu, @r30shah, or @joransiu for the Z codegen related changes here and added documentation.

**Comment by 0xdaryl** [comment]
@genie-omr build all

File: compiler/z/codegen/OMRInstOpCode.hpp [hunk 1]
```diff
@@ -724,6 +724,16 @@ class InstOpCode: public OMR::InstOpCode
...
+   /** \brief
+     *    Gets the 20-bit displacement mnemonic which is equivalent to a 12-bit displacement mnemonic.
+     *
+     *  \param op
+     *     The 12-bit displacement mnemonic for which to find an equivalent 20-bit displacement mnemonic.
+     *
+     *  \return
+     *     The 20-bit displacement mnemonic if it exists, otherwise `BAD` mnemonic.
+     */
+   static Mnemonic getLongDisplacementMnemonic(Mnemonic op);
```
Comment (dchopra001) [inline compiler/z/codegen/OMRInstOpCode.hpp]: Small nit but do you think `getEquivalentLongDisplacementMnemonic` or `getLongDisplacementEquivalent` would be a better name here? I think it makes it a little more clear to a first time reader what the routine intends to do.

Comment (fjeremic) [reply compiler/z/codegen/OMRInstOpCode.hpp]: I agree. Fixed in 66fe5c9

File: compiler/codegen/OMRCodeGenerator.hpp [hunk 2]
```diff
@@ -1143,17 +1136,21 @@ class OMR_EXTENSIBLE CodeGenerator
...
+
+   public:
+
    int32_t internalControlFlowNestingDepth() {return _internalControlFlowNestingDepth;}
    int32_t internalControlFlowSafeNestingDepth() { return _internalControlFlowSafeNestingDepth; }
    void incInternalControlFlowNestingDepth() {_internalControlFlowNestingDepth++;}
    void decInternalControlFlowNestingDepth() {_internalControlFlowNestingDepth--;}
    bool insideInternalControlFlow() {return (_internalControlFlowNestingDepth > _internalControlFlowSafeNestingDepth);}
    void setInternalControlFlowNestingDepth(int32_t depth) { _internalControlFlowNestingDepth = depth; }
    void setInternalControlFlowSafeNestingDepth(int32_t safeDepth) { _internalControlFlowSafeNestingDepth = safeDepth; }
```
Comment (dchopra001) [inline compiler/codegen/OMRCodeGenerator.hpp]: Not sure if this is the right PR for these, but more naming nitpicks:

`internalControlFlowNestingDepth()` can be changed to `getInternalControlFlowNestingDepth` and `internalControlFlowSafeNestingDepth()` can be changed to `getInternalControlFlowSafeNestingDepth()`. This way the API naming convention will be consisted with how the setters are named. 

We can also change `insideInternalControlFlow()` to `isInsideInternalControlFlow()`.

Comment (fjeremic) [reply compiler/codegen/OMRCodeGenerator.hpp]: I think this is an issue outside the scope of this PR as these are common APIs used across other codegens as well. We can however open an issue for this if you think it merits additional work?

Comment (dchopra001) [reply compiler/codegen/OMRCodeGenerator.hpp]: Agreed. I've opened an issue here: https://github.com/eclipse/omr/issues/3373

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by 0xdaryl** [comment]
@genie-omr build xlinux

All xLinux jobs aborted due to connectivity issues.

**Comment by 0xdaryl** [comment]
Actually, @dchopra001 please give this your official blessing please before I merge.

**Comment by 0xdaryl** [comment]
Ping @dchopra001

**Comment by dchopra001** [comment]
LGTM.

**Comment by 0xdaryl** [comment]
@genie-omr build xlinux

One last try for the x86 jobs so I can sleep peacefully with visions of sugarplums dancing in my head.

**Comment by 0xdaryl** [comment]
I'm reasonably confident in the amount of testing that has occurred on x86 for this PR.  Given that only a small part of this PR affects common code and that those changes are relatively straightforward I'm confident in merging this despite the infrastructure failures.

**Pull request merged**